official/resnet/resnet.py

```diff
@@ -169,48 +175,122 @@ def building_block(inputs, filters, training, projection_shortcut, strides,
     The output tensor of the block.
   """
   shortcut = inputs
-  inputs = batch_norm_relu(inputs, training, data_format)
 
-  # The projection shortcut should come after the first batch norm and ReLU
-  # since it performs a 1x1 convolution.
   if projection_shortcut is not None:
     shortcut = projection_shortcut(inputs)
+    shortcut = batch_norm_relu(shortcut, training, data_format)
```

Comment:
I think you don't want this here. Just add `shortcut` to `inputs` below and then call `batch_norm_relu` afterward on the combined value. In this case you can also remove `intermediate_addition` from `batch_norm_relu` above.

Reply:
Ok, if you think the BN-ReLU is unnecessary for the projection I'll remove it. As far as removing the `intermediate_addition` arg, the 2016 v2 paper is clear that the shortcut addition goes between the BN and ReLU for v1, so I don't follow how that can be removed.

Reply:
Looks like the TPU version of v1 just repeats the ReLU, which I suppose also works.

Reply:
I see; you have the right idea with `intermediate_addition` then. In that case, for the current `batch_norm_relu` after this projection shortcut, I think we actually only want a batch norm, since the relu will come after the shortcut and residual are combined together.

The best thing to do then is likely to add a separate function for just batch norm (without relu), which we can use both here and below, in which case we should no longer need the `intermediate_addition` argument.

Reply:
Looks like this is the TPU version you're referring to? https://github.com/tensorflow/tpu/blob/master/models/official/resnet/resnet_model.py#L171

This doesn't look quite correct to me, since there are too many relus. In fact the relu that comes at the end doesn't do anything, since the two values being added together have both been relu'd themselves. Goes along with what I mentioned of every ResNet implementation having minor differences. :)